mini-projects/00-jq-rs/src/filters.rs

```diff
@@ -0,0 +1,156 @@
+use serde_json::Value;
+
+pub fn identity_filter(json: &Value) -> Value {
+    json.clone()
+}
+
+pub fn object_identifier_index(json: &Value, key: &str) -> Value {
+    json.get(key).cloned().unwrap_or(Value::Null)
+}
+
+pub fn array_index(json: &Value, index: usize) -> Value {
+    json.get(index).cloned().unwrap_or(Value::Null)
+}
+
+pub fn array_slice(json: &Value, start: usize, end: usize) -> Value {
+    if let Some(array) = json.as_array() {
+        Value::Array(array[start..end].to_vec())
+    } else {
+        Value::Null
+    }
+}
+
+pub fn pipe(json: &Value, key: &str, index: usize) -> Value {
+    let intermediate = object_identifier_index(json, key);
+    array_index(&intermediate, index)
+}
+
+pub fn array_iterator(json: &Value, key: &str) -> Value {
+    if let Some(array) = json.as_array() {
+        let iterated: Vec<Value> = array
+            .iter()
+            .map(|item| object_identifier_index(item, key))
+            .collect();
+        Value::Array(iterated)
+    } else {
+        Value::Null
+    }
+}
+
+pub fn add(json: &Value) -> Value {
+    if let Some(array) = json.as_array() {
+        let sum = array
+            .iter()
+            .fold(Value::Null, |acc, item| match (acc, item) {
+                (Value::String(mut acc_str), Value::String(item_str)) => {
+                    acc_str.push_str(item_str);
+                    Value::String(acc_str)
+                }
+                (Value::Number(acc_num), Value::Number(item_num)) => {
+                    let acc_f64 = acc_num.as_f64().unwrap_or(0.0);
+                    let item_f64 = item_num.as_f64().unwrap_or(0.0);
+                    Value::Number(serde_json::Number::from_f64(acc_f64 + item_f64).unwrap())
+                }
+                _ => Value::Null,
+            });
+        sum
+    } else {
+        Value::Null
+    }
+}
+
+pub fn length(json: &Value) -> Value {
+    match json {
+        Value::Array(arr) => Value::Number(arr.len().into()),
+        Value::Object(obj) => Value::Number(obj.len().into()),
+        Value::String(s) => Value::Number(s.len().into()),
+        _ => Value::Null,
+    }
+}
+
+pub fn del(json: &mut Value, key_or_index: &str) -> Value {
+    match json {
+        Value::Object(ref mut obj) => {
+            obj.remove(key_or_index);
+            Value::Object(obj.clone())
+        }
+        Value::Array(ref mut arr) => {
+            if let Ok(index) = key_or_index.parse::<usize>() {
+                if index < arr.len() {
+                    arr.remove(index);
+                }
+            }
+            Value::Array(arr.clone())
+        }
+        _ => Value::Null,
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I'd like to see more breadth in your test cases - in particular, only a single positive case is being nested; more could be done to test potential failures/edge cases